K2Bridge/Visitors/AggregationVisitor.cs

```diff
@@ -21,19 +21,28 @@ public void Visit(AggregationContainer aggregationContainer)
                 return;
             }
 
-            aggregationContainer.PrimaryAggregation.Accept(this);
-
             // TODO: do something with the sub aggregations to KQL
             if (aggregationContainer.SubAggregations != null)
             {
                 foreach (var (_, subAgg) in aggregationContainer.SubAggregations)
                 {
                     subAgg.Accept(this);
-                    aggregationContainer.KustoQL += $"{subAgg.KustoQL}, "; // this won't work when 2+ bucket aggregations are used!
+                    if (aggregationContainer.PrimaryAggregation.GetType().BaseType.Name == "BucketAggregation")
+                    {
+                        // Bucket aggregations are responsible for inserting the metrics expressions in the final query
+                        ((BucketAggregation)aggregationContainer.PrimaryAggregation).MetricsKustoQL += $"{subAgg.KustoQL}, ";
+                    }
+                    else
+                    {
+                        // Metrics aggregations just return their KustoQL
+                        aggregationContainer.KustoQL += $"{subAgg.KustoQL}, "; // this won't work when 2+ bucket aggregations are used!
+                    }
                 }
             }
 
-            aggregationContainer.KustoQL += aggregationContainer.PrimaryAggregation.KustoQL;
+            aggregationContainer.PrimaryAggregation.Accept(this);
```

Comment:
Why did you change the order?

Reply:
The Range visitor needs to have all the Metrics KustoQL strings, so that it can use them to assemble its query. So, we first go through all the sub-aggregations (metrics) to gather their KustoQL, and then call the primary aggregation visitor (the bucket aggregation) so that it produces the final query, including the metrics. I have added more comments to explain the flow in this Visitor.